### PR TITLE
Fix bug about multi gpu in LSTM

### PR DESCRIPTION
**Pull request by kamo-naoyuki**
Fix of the issue described here https://github.com/pfnet/chainer/issues/1869.

Reopen from https://github.com/pfnet/chainer/pull/1876(Sorry, this previous PR has some problems and I closed it.)

### CODE REVIEW

**Comment by okuta** [comment]
Could you add tests?

**Comment by zori** [comment]
I updated the PR description to include a link to the issue (https://github.com/pfnet/chainer/issues/1869.)

**Comment by kamo-naoyuki** [comment]
What do you mean "tests"? Do you mean the tests for these links on multi-gpus?
Sorry, I'm not familiar to your test codes.

**Comment by kamo-naoyuki** [comment]
I'll write the omitted explanation of this bug.
 
By default, cupy selects device number 0. so the following array is allocated on device 0.
```
a = cupy.zeros(4)
assert a.device == cuda.get_device(0)
```

If you want to allocated on specified device, you need to use the get_device function with "with" sentence.

```
with with cuda.get_device(1):
  a = cupy.zeros(4)
assert a.device == cuda.get_device(1)
```
Or,

```
cuda.Device(1).use()
a = cupy.zeros(4)
assert a.device == cuda.get_device(1)
```

For example,  the array in https://github.com/pfnet/chainer/blob/285e6558640425f61b6aa8c00564ccf37643babb/chainer/links/connection/lstm.py of Line 220, was allocated on the default device independently of the input array.

I think this is not good specification because many users are note aware that cupy has a default device number and it can be changed by device.use method. In general global variable is not a good idea.
 
There is a another problem. Your chainer.Link class has xp property, so many people might think that an array which was allocated just  by using this module will works well, but really the proper device needs to be specified as following,

```
with cuda.get_device(x.data):
  self.xp.zeros(4)
```

Furthermore, I have another problem. If you'll define new chainer function, you don't have to do as it because Function.__call__ set the device before calling the forward method at line198 of https://github.com/pfnet/chainer/blob/master/chainer/function.py . I think this is a good solution locally but confusing on the whole.

**Comment by okuta** [comment]
https://github.com/kamo-naoyuki/chainer/pull/1

File: chainer/links/connection/lstm.py [hunk 1]
```diff
@@ -91,9 +92,10 @@ def __call__(self, c, h, x):
             lstm_in += self.lateral(h)
         if c is None:
             xp = self.xp
-            c = variable.Variable(
-                xp.zeros((x.shape[0], self.state_size), dtype=x.dtype),
-                volatile='auto')
+            with cuda.get_device(x.data):
```
Comment (okuta) [inline chainer/links/connection/lstm.py]: Please use `with cuda.get_device(self._device_id):`.

File: chainer/links/connection/peephole.py [hunk 2]
```diff
@@ -98,9 +99,10 @@ def __call__(self, x):
             lstm_in += self.lateral(self.h)
         if self.c is None:
             xp = self.xp
-            self.c = variable.Variable(
-                xp.zeros((x.shape[0], self.state_size), dtype=x.dtype),
-                volatile='auto')
+            with cuda.get_device(x.data):
```
Comment (okuta) [inline chainer/links/connection/peephole.py]: Please use `with cuda.get_device(self._device_id):`.

File: chainer/links/connection/lstm.py [hunk 3]
```diff
@@ -239,9 +241,10 @@ def __call__(self, x):
                 lstm_in += self.lateral(self.h)
         if self.c is None:
             xp = self.xp
-            self.c = variable.Variable(
-                xp.zeros((batch, self.state_size), dtype=x.dtype),
-                volatile='auto')
+            with cuda.get_device(x.data):
```
Comment (okuta) [inline chainer/links/connection/lstm.py]: Please use `with cuda.get_device(self._device_id):`.

File: chainer/links/connection/zoneoutlstm.py [hunk 4]
```diff
@@ -86,14 +87,18 @@ def __call__(self, x):
             lstm_in += self.lateral(self.h)
         else:
             xp = self.xp
-            self.h = variable.Variable(
-                xp.zeros((len(x.data), self.state_size), dtype=x.data.dtype),
-                volatile='auto')
+            with cuda.get_device(x.data):
```
Comment (okuta) [inline chainer/links/connection/zoneoutlstm.py]: Please use `with cuda.get_device(self._device_id):`.

File: chainer/links/connection/zoneoutlstm.py [hunk 5]
```diff
@@ -86,14 +87,18 @@ def __call__(self, x):
...
+                self.h = variable.Variable(
+                    xp.zeros((len(x.data), self.state_size),
+                             dtype=x.data.dtype),
+                    volatile='auto')
         if self.c is None:
             xp = self.xp
-            self.c = variable.Variable(
-                xp.zeros((len(x.data), self.state_size), dtype=x.data.dtype),
-                volatile='auto')
+            with cuda.get_device(x.data):
```
Comment (okuta) [inline chainer/links/connection/zoneoutlstm.py]: Please use `with cuda.get_device(self._device_id):`.

File: chainer/links/normalization/batch_normalization.py [hunk 6]
```diff
@@ -106,13 +107,15 @@ def __call__(self, x, test=False, finetune=False):
         if hasattr(self, 'gamma'):
             gamma = self.gamma
         else:
-            gamma = variable.Variable(self.xp.ones(
-                self.avg_mean.shape, dtype=x.dtype), volatile='auto')
+            with cuda.get_device(x.data):
```
Comment (okuta) [inline chainer/links/normalization/batch_normalization.py]: Please use `with cuda.get_device(self._device_id):`.

File: chainer/links/normalization/batch_normalization.py [hunk 7]
```diff
@@ -106,13 +107,15 @@ def __call__(self, x, test=False, finetune=False):
...
+                gamma = variable.Variable(self.xp.ones(
+                    self.avg_mean.shape, dtype=x.dtype), volatile='auto')
         if hasattr(self, 'beta'):
             beta = self.beta
         else:
-            beta = variable.Variable(self.xp.zeros(
-                self.avg_mean.shape, dtype=x.dtype), volatile='auto')
+            with cuda.get_device(x.data):
```
Comment (okuta) [inline chainer/links/normalization/batch_normalization.py]: Please use `with cuda.get_device(self._device_id):`.

**Comment by okuta** [comment]
@kamo-naoyuki: Do you have a time for this PR?

**Comment by kamo-naoyuki** [comment]
@okuta : Sorry for the late  response, I merged your PR.

**Comment by okuta** [comment]
Thank you for merging.

LGTM expect comments.
Please use with `cuda.get_device(self._device_id):`.

**Comment by okuta** [comment]
@kamo-naoyuki: This bug is a critical for multi gpu user.
If you have no time, I merge this PR. After that, I will small problem.
Which is better?

**Comment by kamo-naoyuki** [comment]
Sorry, I'm not used to Github and I don't know what you mean. This PR  can not be merged?

**Comment by okuta** [comment]
Please use with `cuda.get_device(self._device_id):` instead of `with cuda.get_device(x.data):`.

https://github.com/pfnet/chainer/pull/1882#pullrequestreview-8638217

**Comment by kamo-naoyuki** [comment]
It seems already done by your PR. What should i do?

**Comment by okuta** [comment]
My PR fixed problem partially.
The fixing is not finished in `peephole.py`, `zoneoutlstm.py` and `batch_normalization.py`.
https://github.com/pfnet/chainer/pull/1882/files#diff-03a82c30057b5828d72241167e757938R102

**Comment by kamo-naoyuki** [comment]
Sorry, I was not aware of it. So, I think chainer.links.lstm.LSTM.set_state has a same problem. Do you have any idea for it?

**Comment by okuta** [comment]
Thank you for your fixing.

How about `c_.to_gpu(self._device_id)` at [this line](https://github.com/pfnet/chainer/blob/4df382824b6ff95a012643b7aacbc9ad3f738887/chainer/links/connection/zoneoutlstm.py#L61)?

**Comment by okuta** [comment]
Thank you very much.
LGTM!

**Pull request merged**